In [203]:
import pandas as pd
import numpy as np

In [204]:
df = pd.read_csv(r'C:\Users\cleme\laboratorios\PROYECTO-SHARKS-CARLOS\data\attacks.csv', encoding= "ISO-8859-1")

In [205]:
df.head(20)


Case Number                  Date    Year          Type       Country  \
0     2018.06.25           25-Jun-2018  2018.0       Boating           USA   
1     2018.06.18           18-Jun-2018  2018.0    Unprovoked           USA   
2     2018.06.09           09-Jun-2018  2018.0       Invalid           USA   
3     2018.06.08           08-Jun-2018  2018.0    Unprovoked     AUSTRALIA   
4     2018.06.04           04-Jun-2018  2018.0      Provoked        MEXICO   
5   2018.06.03.b           03-Jun-2018  2018.0    Unprovoked     AUSTRALIA   
6   2018.06.03.a           03-Jun-2018  2018.0    Unprovoked        BRAZIL   
7     2018.05.27           27-May-2018  2018.0    Unprovoked           USA   
8   2018.05.26.b           26-May-2018  2018.0    Unprovoked           USA   
9   2018.05.26.a           26-May-2018  2018.0    Unprovoked           USA   
10    2018.05.24           24-May-2018  2018.0      Provoked     AUSTRALIA   
11    2018.05.21           21-May-2018  2018.0    Unprovoked           USA   
12  2018.05.13.b           13-May-2018  2018.0    Unprovoked           USA   
13  2018.05.13.a           13-May-2018  2018.0       Invalid       ENGLAND   
14    2018.05.00              May 2018  2018.0      Provoked     AUSTRALIA   
15    2018.05.12           12-May-2018  2018.0    Unprovoked  SOUTH AFRICA   
16    2018.05.09           09-May-2018  2018.0  Questionable     AUSTRALIA   
17  2018.04.30.R  Reported 30-Apr-2018  2018.0    Unprovoked      THAILAND   
18  2018.04.28.b           28-Apr-2018  2018.0    Unprovoked    COSTA RICA   
19  2018.04.28.a           28-Apr-2018  2018.0    Unprovoked     AUSTRALIA   

                     Area                                           Location  \
0              California                        Oceanside, San Diego County   
1                 Georgia                     St. Simon Island, Glynn County   
2                  Hawaii                                       Habush, Oahu   
3         New South Wales                                 Arrawarra Headland   
4                  Colima                                           La Ticla   
5         New South Wales                                 Flat Rock, Ballina   
6              Pernambuco                              Piedade Beach, Recife   
7                 Florida  Lighhouse Point Park, Ponce Inlet, Volusia County   
8                 Florida                       Cocoa Beach, Brevard  County   
9                 Florida                      Daytona Beach, Volusia County   
10             Queensland                                    Cairns Aquarium   
11         South Carolina                  Isle of Palms,  Charleston County   
12         South Carolina                Hilton Head Island, Beaufort County   
13               Cornwall                                     Off Land's End   
14      Westerm Australia                                         Dugong Bay   
15  Eastern Cape Province                          Nahoon Beach, East London   
16        New South Wales                             Sharpes Beach, Ballina   
17                Hua Hin                                      Sai Noi Beach   
18           Cocos Island                                          Manuelita   
19      Western Australia                                       Gearys Beach   

           Activity                    Name Sex   ...  \
0          Paddling             Julie Wolfe    F  ...   
1          Standing         Adyson McNeely     F  ...   
2           Surfing             John Denges    M  ...   
3           Surfing                    male    M  ...   
4       Free diving          Gustavo Ramos     M  ...   
5      Kite surfing                 Chris 
    M  ...   
6          Swimming  Jose Ernesto da Silva     M  ...   
7           Fishing                    male    M  ...   
8           Walking               Cody High    M  ...   
9          Standing                    male    M  ...   
10   Feeding sharks                    male    M  ...   
1

In [206]:
df.shape

(25723, 24)

In [207]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

  1 Paso.- Consiste en filtrar todas las filas que contengan un número de celdas sin información en un número de 8 o másy eliminarlas considerando que la tabla tiene 23 cifras. 

In [208]:
df = df[df.count(axis=1) >= 8]


In [209]:
df.shape

(6302, 24)

2 Paso. Rellenar todas las celdas sin datos con 'unknow', considerando como 8 celdas en la fila la menor información posible para ser interpretable.

In [210]:
df = df.fillna('unknown')

In [211]:
df


Case Number         Date    Year        Type             Country  \
0     2018.06.25  25-Jun-2018  2018.0     Boating                 USA   
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked                 USA   
2     2018.06.09  09-Jun-2018  2018.0     Invalid                 USA   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   
4     2018.06.04  04-Jun-2018  2018.0    Provoked              MEXICO   
...          ...          ...     ...         ...                 ...   
6297     ND.0005  Before 1903     0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903     0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905     0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889     0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   Area                             Location      Activity  \
0            California          Oceanside, San Diego County      Paddling   
1               Georgia       St. Simon Island, Glynn County      Standing   
2                Hawaii                         Habush, Oahu       Surfing   
3       New South Wales                   Arrawarra Headland       Surfing   
4                Colima                             La Ticla   Free diving   
...                 ...                                  ...           ...   
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                              unknown  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300            unknown                 Panama Bay 8ºN, 79ºW       unknown   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       Name Sex   ...         Species   \
0               Julie Wolfe    F  ...      White shark   
1           Adyson McNeely     F  ...          unknown   
2               John Denges    M  ...          unknown   
3                      male    M  ...        2 m shark   
4            Gustavo Ramos     M  ...  Tiger shark, 3m   
...                     ...  ...  ...              ...   
6297                   male    M  ...          unknown   
6298                  Ahmun    M  ...          unknown   
6299  Coast Guard personnel    M  ...          unknown   
6300        Jules Patterson    M  ...          unknown   
6301                   male    M  ...          unknown   

                     Investigator or Source                             pdf  \
0                          R. Collier, GSAF            2018.06.25-Wolfe.pdf   
1            K.McMurray, TrackingSharks.com          2018.06.18-McNeely.pdf   
2            K.McMurray, TrackingSharks.com           2018.06.09-Denges.pdf   
3                            B. Myatt, GSAF        2018.06.08-Arrawarra.pdf   
4                                 A .Kipper            2018.06.04-Ramos.pdf   
...                                     ...                             ...   
6297       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
6298  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
6299   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
6300                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
6301                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href formula  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...   
1     http://sharkattackfile.net/spreadsheets/pdf_di...   
2     http://sharkattackfile.net/spreadsheets/pdf_di...   
3     http://sharkattackfile.net/spreadsheets/pdf_di...   
4     http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                 ...   
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://s

3 paso Elimino la fila de año para rellenarla con el índice de 1 a n registros, en definitiva se denomina "Case number". Ejecuto un bucle  que me cambia esta columna de forma inversa del 1 en la última posición hasta el primero en la posición 6302


In [212]:

n_fila = df.shape[0]

for i in range(n_fila):
    df.loc[i, 'Case Number'] = n_fila - i
    

df


Case Number         Date    Year        Type             Country  \
0           6302  25-Jun-2018  2018.0     Boating                 USA   
1           6301  18-Jun-2018  2018.0  Unprovoked                 USA   
2           6300  09-Jun-2018  2018.0     Invalid                 USA   
3           6299  08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   
4           6298  04-Jun-2018  2018.0    Provoked              MEXICO   
...          ...          ...     ...         ...                 ...   
6297           5  Before 1903     0.0  Unprovoked           AUSTRALIA   
6298           4  Before 1903     0.0  Unprovoked           AUSTRALIA   
6299           3    1900-1905     0.0  Unprovoked                 USA   
6300           2    1883-1889     0.0  Unprovoked              PANAMA   
6301           1    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   Area                             Location      Activity  \
0            California          Oceanside, San Diego County      Paddling   
1               Georgia       St. Simon Island, Glynn County      Standing   
2                Hawaii                         Habush, Oahu       Surfing   
3       New South Wales                   Arrawarra Headland       Surfing   
4                Colima                             La Ticla   Free diving   
...                 ...                                  ...           ...   
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                              unknown  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300            unknown                 Panama Bay 8ºN, 79ºW       unknown   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       Name Sex   ...         Species   \
0               Julie Wolfe    F  ...      White shark   
1           Adyson McNeely     F  ...          unknown   
2               John Denges    M  ...          unknown   
3                      male    M  ...        2 m shark   
4            Gustavo Ramos     M  ...  Tiger shark, 3m   
...                     ...  ...  ...              ...   
6297                   male    M  ...          unknown   
6298                  Ahmun    M  ...          unknown   
6299  Coast Guard personnel    M  ...          unknown   
6300        Jules Patterson    M  ...          unknown   
6301                   male    M  ...          unknown   

                     Investigator or Source                             pdf  \
0                          R. Collier, GSAF            2018.06.25-Wolfe.pdf   
1            K.McMurray, TrackingSharks.com          2018.06.18-McNeely.pdf   
2            K.McMurray, TrackingSharks.com           2018.06.09-Denges.pdf   
3                            B. Myatt, GSAF        2018.06.08-Arrawarra.pdf   
4                                 A .Kipper            2018.06.04-Ramos.pdf   
...                                     ...                             ...   
6297       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
6298  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
6299   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
6300                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
6301                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href formula  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...   
1     http://sharkattackfile.net/spreadsheets/pdf_di...   
2     http://sharkattackfile.net/spreadsheets/pdf_di...   
3     http://sharkattackfile.net/spreadsheets/pdf_di...   
4     http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                 ...   
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://s